README.md
Code uses PyTorch as the core library for the training. Heavily inspired by their finetuning tutorial page. 
This project uses the Deep-CNN Inception V3 to classify chest x-ray images from the Kaggle dataset covid19-image-dataset (uploaded by pranavraikokte). 
CUDA toolkit can be used (if available), for significantly faster processing in both training and evaluation. 

The dataset consists of 3 different types of x-rays, 1. chest with healthy lungs, 2. chest with viral infection on the lungs and 3. chest with sars cov 2 infection on the lungs.
The network_trainer.py file trains the CNN on these images for classification on each of the three types of lungs. 
Mulpiple approaches of training are presented, such as training of the whole network or training of the final layer, in the case of choosing to have pretrained values (from ImageNet) loaded into the network.
Images are resized to 299x299 pixels and randomly editted (resized and rotated). The values of the images are also normalized. 
Since the final layer of the network has 1024 classes, we change that to our corresponding 3 lung types.

Best results were presented with the fine tuning training process. 

The model_tester.py file can be used to load saved networks and use them for classification testing. 
(Network must be same type as Inception V3, or the code has to be tweaked to match the new network's structure)

Link for dataset: https://www.kaggle.com/datasets/pranavraikokte/covid19-image-dataset

Link for PyTorch tutorial: https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html


history.txt
Total Epochs: 50
Batch Size: 8
Device: cuda:0
Pretrained: False
Model: inception
Epoch: 1 Test Accuracy: 0.3030 
Epoch: 2 Test Accuracy: 0.3030 
Epoch: 3 Test Accuracy: 0.4545 
Epoch: 4 Test Accuracy: 0.4848 
Epoch: 5 Test Accuracy: 0.5606 
Epoch: 6 Test Accuracy: 0.6061 
Epoch: 7 Test Accuracy: 0.6818 
Epoch: 8 Test Accuracy: 0.5606 
Epoch: 9 Test Accuracy: 0.6818 
Epoch: 10 Test Accuracy: 0.6970 
Epoch: 11 Test Accuracy: 0.6515 
Epoch: 12 Test Accuracy: 0.4242 
Epoch: 13 Test Accuracy: 0.6970 
Epoch: 14 Test Accuracy: 0.7576 
Epoch: 15 Test Accuracy: 0.6970 
Epoch: 16 Test Accuracy: 0.6212 
Epoch: 17 Test Accuracy: 0.8333 
Epoch: 18 Test Accuracy: 0.7121 
Epoch: 19 Test Accuracy: 0.6818 
Epoch: 20 Test Accuracy: 0.7424 
Epoch: 21 Test Accuracy: 0.8636 
Epoch: 22 Test Accuracy: 0.6970 
Epoch: 23 Test Accuracy: 0.6818 
Epoch: 24 Test Accuracy: 0.6970 
Epoch: 25 Test Accuracy: 0.7121 
Epoch: 26 Test Accuracy: 0.6970 
Epoch: 27 Test Accuracy: 0.7273 
Epoch: 28 Test Accuracy: 0.7273 
Epoch: 29 Test Accuracy: 0.8030 
Epoch: 30 Test Accuracy: 0.7727 
Epoch: 31 Test Accuracy: 0.8030 
Epoch: 32 Test Accuracy: 0.6667 
Epoch: 33 Test Accuracy: 0.6970 
Epoch: 34 Test Accuracy: 0.6364 
Epoch: 35 Test Accuracy: 0.8485 
Epoch: 36 Test Accuracy: 0.8485 
Epoch: 37 Test Accuracy: 0.6364 
Epoch: 38 Test Accuracy: 0.8939 
Epoch: 39 Test Accuracy: 0.7121 
Epoch: 40 Test Accuracy: 0.8333 
Epoch: 41 Test Accuracy: 0.8485 
Epoch: 42 Test Accuracy: 0.8636 
Epoch: 43 Test Accuracy: 0.8636 
Epoch: 44 Test Accuracy: 0.7273 
Epoch: 45 Test Accuracy: 0.8333 
Epoch: 46 Test Accuracy: 0.8333 
Epoch: 47 Test Accuracy: 0.7576 
Epoch: 48 Test Accuracy: 0.8636 
Epoch: 49 Test Accuracy: 0.7273 
Epoch: 50 Test Accuracy: 0.9091 

###############################
Total Epochs: 50
Batch Size: 8
Device: cuda:0
Pretrained: True
Model: inception
Epoch: 1 Test Accuracy: 0.8030 
Epoch: 2 Test Accuracy: 0.8030 
Epoch: 3 Test Accuracy: 0.8485 
Epoch: 4 Test Accuracy: 0.7879 
Epoch: 5 Test Accuracy: 0.9545 
Epoch: 6 Test Accuracy: 0.9697 
Epoch: 7 Test Accuracy: 0.8485 
Epoch: 8 Test Accuracy: 0.9394 
Epoch: 9 Test Accuracy: 0.9545 
Epoch: 10 Test Accuracy: 0.9848 
Epoch: 11 Test Accuracy: 0.8485 
Epoch: 12 Test Accuracy: 0.9697 
Epoch: 13 Test Accuracy: 0.9848 
Epoch: 14 Test Accuracy: 0.8939 
Epoch: 15 Test Accuracy: 1.0000 
Epoch: 16 Test Accuracy: 0.9848 
Epoch: 17 Test Accuracy: 0.9545 
Epoch: 18 Test Accuracy: 0.9848 
Epoch: 19 Test Accuracy: 1.0000 
Epoch: 20 Test Accuracy: 0.9697 
Epoch: 21 Test Accuracy: 0.9697 
Epoch: 22 Test Accuracy: 1.0000 
Epoch: 23 Test Accuracy: 0.9394 
Epoch: 24 Test Accuracy: 0.8182 
Epoch: 25 Test Accuracy: 0.9848 
Epoch: 26 Test Accuracy: 0.9848 
Epoch: 27 Test Accuracy: 0.9697 
Epoch: 28 Test Accuracy: 0.9091 
Epoch: 29 Test Accuracy: 0.9545 
Epoch: 30 Test Accuracy: 0.9545 
Epoch: 31 Test Accuracy: 0.9848 
Epoch: 32 Test Accuracy: 0.9091 
Epoch: 33 Test Accuracy: 0.9394 
Epoch: 34 Test Accuracy: 0.9394 
Epoch: 35 Test Accuracy: 0.9242 
Epoch: 36 Test Accuracy: 0.9697 
Epoch: 37 Test Accuracy: 0.9242 
Epoch: 38 Test Accuracy: 0.9545 
Epoch: 39 Test Accuracy: 0.9545 
Epoch: 40 Test Accuracy: 0.9848 
Epoch: 41 Test Accuracy: 0.9848 
Epoch: 42 Test Accuracy: 0.9545 
Epoch: 43 Test Accuracy: 0.9848 
Epoch: 44 Test Accuracy: 0.9848 
Epoch: 45 Test Accuracy: 1.0000 
Epoch: 46 Test Accuracy: 0.9242 
Epoch: 47 Test Accuracy: 0.9697 
Epoch: 48 Test Accuracy: 0.9697 
Epoch: 49 Test Accuracy: 0.8939 
Epoch: 50 Test Accuracy: 1.0000 

###############################
Feature Extract: True
Total Epochs: 50
Batch Size: 8
Device: cuda:0
Pretrained: True
Model: inception
Epoch: 1 Test Accuracy: 0.4545 
Epoch: 2 Test Accuracy: 0.6515 
Epoch: 3 Test Accuracy: 0.7121 
Epoch: 4 Test Accuracy: 0.7121 
Epoch: 5 Test Accuracy: 0.8485 
Epoch: 6 Test Accuracy: 0.8030 
Epoch: 7 Test Accuracy: 0.8182 
Epoch: 8 Test Accuracy: 0.7727 
Epoch: 9 Test Accuracy: 0.7576 
Epoch: 10 Test Accuracy: 0.7727 
Epoch: 11 Test Accuracy: 0.7424 
Epoch: 12 Test Accuracy: 0.7879 
Epoch: 13 Test Accuracy: 0.8485 
Epoch: 14 Test Accuracy: 0.8636 
Epoch: 15 Test Accuracy: 0.8182 
Epoch: 16 Test Accuracy: 0.8485 
Epoch: 17 Test Accuracy: 0.8636 
Epoch: 18 Test Accuracy: 0.8939 
Epoch: 19 Test Accuracy: 0.8636 
Epoch: 20 Test Accuracy: 0.8333 
Epoch: 21 Test Accuracy: 0.8788 
Epoch: 22 Test Accuracy: 0.9091 
Epoch: 23 Test Accuracy: 0.8636 
Epoch: 24 Test Accuracy: 0.8333 
Epoch: 25 Test Accuracy: 0.8485 
Epoch: 26 Test Accuracy: 0.8333 
Epoch: 27 Test Accuracy: 0.8182 
Epoch: 28 Test Accuracy: 0.8636 
Epoch: 29 Test Accuracy: 0.8333 
Epoch: 30 Test Accuracy: 0.8333 
Epoch: 31 Test Accuracy: 0.8333 
Epoch: 32 Test Accuracy: 0.8636 
Epoch: 33 Test Accuracy: 0.8788 
Epoch: 34 Test Accuracy: 0.8939 
Epoch: 35 Test Accuracy: 0.8939 
Epoch: 36 Test Accuracy: 0.8485 
Epoch: 37 Test Accuracy: 0.8333 
Epoch: 38 Test Accuracy: 0.8939 
Epoch: 39 Test Accuracy: 0.8030 
Epoch: 40 Test Accuracy: 0.8485 
Epoch: 41 Test Accuracy: 0.8636 
Epoch: 42 Test Accuracy: 0.8636 
Epoch: 43 Test Accuracy: 0.8485 
Epoch: 44 Test Accuracy: 0.8788 
Epoch: 45 Test Accuracy: 0.8333 
Epoch: 46 Test Accuracy: 0.8939 
Epoch: 47 Test Accuracy: 0.8030 
Epoch: 48 Test Accuracy: 0.8182 
Epoch: 49 Test Accuracy: 0.8333 
Epoch: 50 Test Accuracy: 0.8788 


Code/model_tester.py
from __future__ import print_function
from __future__ import division
import torch
import torch.nn as nn
from torchvision import datasets, models, transforms
import time
import os

# Give model's file path
model_path = 'C:/Users/simos/Desktop/Covid19-dataset/saved_models/inceptionv3covid-32-50-Pretrain.pt'
# Give model's name
model_name = 'inception'
# Give model's number of output classes
num_classes = 3
# Give input size for the model
input_size = 299
# Give the directory of the dataset
data_dir = 'C:/Users/simos/Desktop/Covid19-dataset/'
# Give a batch size
batch_size = 32
# Set device to cuda if available
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

def initialize_model(model_name, model_path):
    """
    Initialize the model by loading the saved model file into it
    """
    # Initialize these variables which will be set in this if statement. Each of these
    #   variables is model specific.
    model_ft = None
    input_size = 0

    if model_name == "inception":
        """ Inception v3
        Be careful, expects (299,299) sized images and has auxiliary output
        """
        # model_ft = models.inception_v3(pretrained=use_pretrained)
        # # set_parameter_requires_grad(model_ft, feature_extract)
        # # Handle the auxilary net
        # num_ftrs = model_ft.AuxLogits.fc.in_features
        # model_ft.AuxLogits.fc = nn.Linear(num_ftrs, num_classes)
        # # Handle the primary net
        # num_ftrs = model_ft.fc.in_features
        # model_ft.fc = nn.Linear(num_ftrs, num_classes)
        # input_size = 299
        # model_ft.load_state_dict(torch.load(model_path, torch.device('cpu')))  # Save the network model
        model_ft = torch.load(model_path)
        model_ft.eval()  # Set the network to evaluation mode

    else:
        print("Invalid model name, exiting...")
        exit()

    return model_ft, input_size

# Data normalization for validation
data_transforms = {
    'val': transforms.Compose([
        transforms.Resize(input_size),
        transforms.CenterCrop(input_size),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
}

print("Initializing Datasets and Dataloaders...")

# Create validation dataset
image_datasets = {x: datasets.ImageFolder(os.path.join(data_dir, x), data_transforms[x]) for x in
                  ['val']}
# Create validation dataloader
dataloaders_dict = {
    x: torch.utils.data.DataLoader(image_datasets[x], batch_size=batch_size, shuffle=True, num_workers=0) for x in
    ['val']}

def test_model(model, dataloaders, criterion):
    """
    Evaluate the model by passing desired images into its input.
    This can also be used for actual image classification preditions, since the network model is pretrained.
    """
    since = time.time()

    # Each epoch has a training and validation phase
    model.eval()  # Set model to evaluate mode

    running_loss = 0.0
    running_corrects = 0

    # Iterate over data.
    for inputs, labels in dataloaders['val']:
        with torch.set_grad_enabled(False):
            inputs = inputs.to(device)
            labels = labels.to(device)
            model = model.to(device)
            # Get model outputs and calculate loss
            # Special case for inception because in training it has an auxiliary output.
            # In testing we only consider the final output.
            outputs = model(inputs)
            # print(outputs)
            # print(labels)
            loss = criterion(outputs, labels)

            _, preds = torch.max(outputs, 1)

        # statistics
        running_loss += loss.item() * inputs.size(0)
        running_corrects += torch.sum(preds == labels.data)

    epoch_loss = running_loss / len(dataloaders['val'].dataset)
    epoch_acc = running_corrects.double() / len(dataloaders['val'].dataset)

    print('Validation Loss: {:.4f} Acc: {:.4f}'.format(epoch_loss, epoch_acc))

    print()

    time_elapsed = time.time() - since
    print('Testing complete in {:.0f}m {:.0f}s'.format(time_elapsed // 60, time_elapsed % 60))


# Initialize the model for this run
model_ft, input_size = initialize_model(model_name, model_path)
# Setup the loss fxn
criterion = nn.CrossEntropyLoss()
# Run an evaluation round on the model we trained
test_model(model_ft, dataloaders_dict, criterion)


Code/network_trainer.py
from __future__ import print_function
from __future__ import division
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
from torchvision import datasets, models, transforms
import time
import os
import copy

print("PyTorch Version: ", torch.__version__)
print("Torchvision Version: ", torchvision.__version__)

##############################################################################################
# Top level data directory. Here we assume the format of the directory conforms
#   to the ImageFolder structure
data_dir = 'C:/Users/simos/Desktop/Covid19-dataset/'

# Models to choose from [resnet, alexnet, vgg, squeezenet, densenet, inception]
model_name = "inception"

# Number of classes in the dataset
num_classes = 3

# Batch size for training (change depending on how much memory you have)
batch_size = 8

# Number of epochs to train for
num_epochs = 2

# Use pretrained model in ImageNet
pretrained = True

# Flag for feature extracting. When False, we finetune the whole model,
#   when True we only update the reshaped layer params
feature_extract = False


##############################################################################################
def train_model(model, dataloaders, criterion, optimizer, num_epochs=25, is_inception=True):
    """
    Trains and validates the model, depending on the state of the variable 'phase'. It keeps
    the state of the model with the highest validation score.
    """
    since = time.time()

    val_acc_history = []

    best_model_wts = copy.deepcopy(model.state_dict())
    best_acc = 0.0

    for epoch in range(num_epochs):
        print('Epoch {}/{}'.format(epoch + 1, num_epochs))
        print('-' * 10)

        # Each epoch has a training and validation phase
        for phase in ['train', 'val']:
            if phase == 'train':
                model.train()  # Set model to training mode
            else:
                model.eval()  # Set model to evaluate mode

            running_loss = 0.0
            running_corrects = 0

            # Iterate over data.
            for inputs, labels in dataloaders[phase]:
                inputs = inputs.to(device)
                labels = labels.to(device)

                # zero the parameter gradients
                optimizer.zero_grad()

                # forward
                # track history if only in train
                with torch.set_grad_enabled(phase == 'train'):
                    # Get model outputs and calculate loss
                    # Special case for inception because in training it has an auxiliary output. In train
                    #   mode we calculate the loss by summing the final output and the auxiliary output
                    #   but in testing we only consider the final output.
                    if is_inception and phase == 'train':
                        # From https://discuss.pytorch.org/t/how-to-optimize-inception-model-with-auxiliary-classifiers/7958
                        outputs, aux_outputs = model(inputs)
                        loss1 = criterion(outputs, labels)
                        loss2 = criterion(aux_outputs, labels)
                        loss = loss1 + 0.4 * loss2
                    else:
                        outputs = model(inputs)
                        loss = criterion(outputs, labels)

                    _, preds = torch.max(outputs, 1)

                    # backward + optimize only if in training phase
                    if phase == 'train':
                        loss.backward()
                        optimizer.step()

                # statistics
                running_loss += loss.item() * inputs.size(0)
                running_corrects += torch.sum(preds == labels.data)

            epoch_loss = running_loss / len(dataloaders[phase].dataset)
            epoch_acc = running_corrects.double() / len(dataloaders[phase].dataset)

            print('{} Loss: {:.4f} Acc: {:.4f}'.format(phase, epoch_loss, epoch_acc))

            # deep copy the model
            if phase == 'val' and epoch_acc > best_acc:
                best_acc = epoch_acc
                best_model_wts = copy.deepcopy(model.state_dict())
            if phase == 'val':
                val_acc_history.append(epoch_acc)

        print()

    time_elapsed = time.time() - since
    print('Training complete in {:.0f}m {:.0f}s'.format(time_elapsed // 60, time_elapsed % 60))
    print('Best val Acc: {:4f}'.format(best_acc))

    # load best model weights
    model.load_state_dict(best_model_wts)
    return model, val_acc_history


##############################################################################################
def set_parameter_requires_grad(model, feature_extracting):
    """
    Set requires_grad to True or False in the model's parameters
    """
    if feature_extracting:
        for param in model.parameters():
            param.requires_grad = False


##############################################################################################
def initialize_model(model_name, num_classes, feature_extract, use_pretrained=True):
    """
    Initializes the neural network model, setting up the structure and the type
    """
    # Initialize these variables which will be set in this if statement. Each of these
    #   variables is model specific.
    model_ft = None
    input_size = 0

    if model_name == "inception":
        """ Inception v3
        Be careful, expects (299,299) sized images and has auxiliary output
        """
        model_ft = models.inception_v3(pretrained=use_pretrained)
        set_parameter_requires_grad(model_ft, feature_extract)
        # Handle the auxilary net
        num_ftrs = model_ft.AuxLogits.fc.in_features
        model_ft.AuxLogits.fc = nn.Linear(num_ftrs, num_classes)
        # Handle the primary net
        num_ftrs = model_ft.fc.in_features
        model_ft.fc = nn.Linear(num_ftrs, num_classes)
        input_size = 299

    else:
        print("Invalid model name, exiting...")
        exit()

    return model_ft, input_size


# Initialize the model for this run
model_ft, input_size = initialize_model(model_name, num_classes, feature_extract, pretrained)

# Print the model we just instantiated
print(model_ft)

##############################################################################################
# Data augmentation and normalization for training
# Just normalization for validation
data_transforms = {
    'train': transforms.Compose([
        transforms.RandomResizedCrop(input_size),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
    'val': transforms.Compose([
        transforms.Resize(input_size),
        transforms.CenterCrop(input_size),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
}

##############################################################################################
def databackup(hist, model_ft):
    """
    Writing and saving the history of each training session
    """
    # Save in one file the data and some of the model's properties-hyperparameters
    file = open('history.txt', 'a')
    # Save in one file the raw data
    raw_file = open('raw.txt', 'a')
    file.write("Total Epochs: {}\nBatch Size: {}\nDevice: {}\nPretrained: {}\nModel: {}\n".format(num_epochs,
                                                                                                  batch_size,
                                                                                                  device, pretrained,
                                                                                                  model_name))
    for i, acc in enumerate(hist):
        file.write("Epoch: {} ".format(str(i + 1)) + "Test Accuracy: {:.4f} \n".format(acc.data.item()))
        raw_file.write("{} ".format(acc.data.item()))
    raw_file.write("\n")
    file.write("\n###############################\n")
    file.close()

    # Save the trained model for later use
    torch.save(model_ft, data_dir + 'saved_models/inceptionv3covid.pt')  # Save the network model

##############################################################################################
# Detect if we have a GPU available
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

# Send the model to GPU
model_ft = model_ft.to(device)

##############################################################################################
# Gather the parameters to be optimized/updated in this run. If we are
#  finetuning we will be updating all parameters. However, if we are
#  doing feature extract method, we will only update the parameters
#  that we have just initialized, i.e. the parameters with requires_grad
#  is True.
params_to_update = model_ft.parameters()
print("Params to learn:")
if feature_extract:
    params_to_update = []
    for name, param in model_ft.named_parameters():
        if param.requires_grad == True:
            params_to_update.append(param)
            print("\t", name)
else:
    for name, param in model_ft.named_parameters():
        if param.requires_grad == True:
            print("\t", name)

##############################################################################################
# Observe that all parameters are being optimized
optimizer_ft = optim.SGD(params_to_update, lr=0.001, momentum=0.9)

# Setup the loss fxn
criterion = nn.CrossEntropyLoss()

# def dataset_loader(ratio):
print("Initializing Datasets and Dataloaders...")

# Create training and validation datasets
image_datasets = {x: datasets.ImageFolder(os.path.join(data_dir, x), data_transforms[x]) for x in ['train', 'val']}

# Create training and validation dataloaders
dataloaders_dict = {
    x: torch.utils.data.DataLoader(image_datasets[x], batch_size=batch_size, shuffle=True, num_workers=0) for x in
    ['train', 'val']}

##############################################################################################
# Train and evaluate
model_ft, hist = train_model(model_ft, dataloaders_dict, criterion, optimizer_ft, num_epochs=num_epochs,
                             is_inception=(model_name == "inception"))
# Write down the data we just collected
databackup(hist, model_ft)
